### ISSUE TITLE
KSP resolves classes that are not available to the app

### ISSUE THREAD

**Issue by yigit** [google/ksp#198]
Seems like something is wrong with how classpaths are set for android apps (couldn't repro w/ jvm sample but that might be my mistake).

See the attached project where I have an android module and a ksp processor module.

The KSP processor checks for the availability of the `javax.annotation.processing.Generated` class and adds that annotation if available.
Unfortunately, KSP does return that class even though it is not available to the app, causing compilation failure.

Make sure to compile the attached project with a JDK where that class is available. (I'm using J11). I'm *guessing* it is why the class shows up.

I tried adding the following line to the app module but didn't help:

```
tasks.withType(org.jetbrains.kotlin.gradle.tasks.KotlinCompile.class).configureEach {
    it.kotlinOptions.jvmTarget = "1.8"
}
```

[ksp-android-classpath-bug.zip](https://github.com/google/ksp/files/5719966/ksp-android-classpath-bug.zip)

**Comment by eamonnmcmanus** [comment]
FWIW auto-common changed from trying to tell whether `javax.annotation.processing.Generated` is present to just [checking the source version](https://github.com/google/auto/blob/65600f2e8ba4e905db3b12b7424c6264dbb34427/common/src/main/java/com/google/auto/common/GeneratedAnnotations.java#L54) for what are probably reasons similar to the issue here.

**Comment by yigit** [comment]
interesting, i think we might be using the old version (i checked the impl while copying it).
That being said, i don't think KSP provides us that information right now.

Also, `org.gradle.java.home=JDK_8` is a possible workaround but that is still limited to compiling w/ java8 and most clients are already above that.
related kapt issue: https://youtrack.jetbrains.com/issue/KT-32804

**Comment by yigit** [comment]
i found a workaround by calling setupCompilerArgs on the compile kotlin task in the KspTask.

Kapt uses something similar:
https://github.com/JetBrains/kotlin/blob/master/libraries/tools/kotlin-gradle-plugin/src/main/kotlin/org/jetbrains/kotlin/gradle/tasks/Tasks.kt#L502

but it is using the contributor. I've not debugged what is in it. Specifically, if it adds sources etc, that might be a problem. Otherwise, seems like a safe solution?

**Comment by ting-yuan** [comment]
Fixed in #270. Solution is what Yigit mentioned in the above.
